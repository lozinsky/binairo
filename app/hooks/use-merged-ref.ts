import { type Ref, useMemo } from 'react';

export function useMergedRef<T>(refs: ReadonlyArray<Ref<T> | undefined>) {
  const mergedRef = useMemo<Ref<T>>(() => {
    if (refs.every((ref) => ref == null)) {
      return null;
    }

    return (value) => {
      for (const ref of refs) {
        if (typeof ref === 'function') {
          ref(value);
        } else if (ref != null) {
          ref.current = value;
        }
      }
    };
  }, refs); // eslint-disable-line react-hooks/exhaustive-deps

  return mergedRef;
}
