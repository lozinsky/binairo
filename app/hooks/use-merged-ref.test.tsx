/* @vitest-environment jsdom */

import { renderHook } from '@testing-library/react';
import { useImperativeHandle, useRef } from 'react';
import { expect, test, vi } from 'vitest';

import { useMergedRef } from './use-merged-ref';

test.each([
  () => {
    return [];
  },
  () => {
    return [null, null];
  },
  () => {
    return [undefined, undefined];
  },
  () => {
    return [null, undefined];
  },
])('returns "null" if no refs defined', (useRefs) => {
  const { result } = renderHook(() => {
    const refs = useRefs();
    const mergedRef = useMergedRef(refs);

    useImperativeHandle(mergedRef, () => 'value');

    return { mergedRef, refs };
  });

  expect(result.current.mergedRef).toMatchSnapshot();
});

test.each([
  () => {
    const ref1 = useRef(null);
    const ref2 = useRef(null);

    return [ref1, ref2, null, undefined];
  },
  () => {
    const ref1 = vi.fn();
    const ref2 = vi.fn();

    return [ref1, ref2, null, undefined];
  },
  () => {
    const ref1 = useRef(null);
    const ref2 = vi.fn();

    return [ref1, ref2, null, undefined];
  },
])('returns merged ref', (useRefs) => {
  const { result } = renderHook(() => {
    const refs = useRefs();
    const mergedRef = useMergedRef(refs);

    useImperativeHandle(mergedRef, () => 'value');

    return { mergedRef, refs };
  });

  expect(result.current.refs).toMatchSnapshot();
});
