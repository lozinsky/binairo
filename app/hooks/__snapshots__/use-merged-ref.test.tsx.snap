// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns "null" if no refs defined 1`] = `null`;

exports[`returns "null" if no refs defined 2`] = `null`;

exports[`returns "null" if no refs defined 3`] = `null`;

exports[`returns "null" if no refs defined 4`] = `null`;

exports[`returns merged ref 1`] = `
[
  {
    "current": "value",
  },
  {
    "current": "value",
  },
  null,
  undefined,
]
`;

exports[`returns merged ref 2`] = `
[
  [MockFunction spy] {
    "calls": [
      [
        "value",
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  [MockFunction spy] {
    "calls": [
      [
        "value",
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  null,
  undefined,
]
`;

exports[`returns merged ref 3`] = `
[
  {
    "current": "value",
  },
  [MockFunction spy] {
    "calls": [
      [
        "value",
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  null,
  undefined,
]
`;
